site/_posts/2017-12-13-java-vector-improvements.md

```diff
@@ -0,0 +1,110 @@
+---
+layout: post
+title: "Improved JAVA Vector APIs"
+excerpt: "This post describes the recent improvements in JAVA Vector code"
+date: 2017-12-13 12:50:00
+author: Siddharth Teotia
+categories: [application]
+---
+
+<!--
+{% comment %}
+Licensed to the Apache Software Foundation (ASF) under one or more
+contributor license agreements.  See the NOTICE file distributed with
+this work for additional information regarding copyright ownership.
+The ASF licenses this file to you under the Apache License, Version 2.0
+(the "License"); you may not use this file except in compliance with
+the License.  You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+{% endcomment %}
+-->
+
+
+This post gives insight into the major improvements in the JAVA implementation
+of vectors.
+
+## Design Goals
+
+1. Improved Maintainability and Extensibility.
+2. Improved heap usage.
+3. No performance overhead on hot code paths.
+
+## Background
+
+**Improved Maintainability and Extensibility** 
+
+We use templates in several places for compile time JAVA code generation for
+different vector classes, readers, writers etc. Templates are helpful as the
+developers don't have to write a lot of duplicate code. 
+
+However, we realized that over a period of time some specific JAVA 
+templates became extremely complex with giant if-else blocks, poor code indentation
+and documentation. All this impacted the ability to easily extend these templates 
+for adding new functionality or improving the existing infrastructure.
+
+So we evaluated the usage of templates for compile time code generation and
+decided not to use complex templates in some places by writing small amount of 
+duplicate code which is elegant, well documented and extensible.
+
+**Improved Heap Usage**
+
+We did extensive memory analysis downstream in Dremio where Arrow is used
+heavily for in-memory query execution on columnar data. The general conclusion
+was that Arrow JAVA Vectors have non-negligible heap overhead and volume of 
+objects was too high. There were places in code where we were creating objects
+unnecessarily and using structures that could be substituted with better
+alternatives.
+
+**No performance overhead on hot code paths**
```

Comment:
The other two section  "Improved Maintainability and Extensibility" and "Improved Heap Usage" are both capitalized, maybe make them more consistent?

Reply:
Done.